### PR TITLE
RavenDB-19168: Optimizations

### PR DESCRIPTION
**Pull request by redknightlois**
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-19168
https://issues.hibernatingrhinos.com/issue/RavenDB-19190
https://issues.hibernatingrhinos.com/issue/RavenDB-16837
https://issues.hibernatingrhinos.com/issue/RavenDB-18158/Optimize-DynamicArray

### Additional description

Multiple optimizations and simplifications at the low level code. Standarizing usage of ZigZagEncoding and VariableLengthEncoding, avoid castings, etc.

### Type of change

- Optimization

### How risky is the change?

- Low

### CODE REVIEW

**Comment by arekpalinski** [comment]
The scope of optimizations made here is wider than Corax only. Changes in blittables / dynamic arrays worry me. I wonder if those shouldn't go to v6.0. 

@redknightlois @ayende what's your opinion on that?

**Comment by redknightlois** [comment]
We could send those to 6.0 but they are not such complicated changes. In
fact I killed one that is too risky for 5.4. They are really hot paths, if
those are wrong they would trip tests all over the place, that's why I set
risk as low. But I understand if we don't want to risk. Your call.

**Comment by ayende** [comment]
Not for the next stable, but for the nightlies, should be enough.
The point is that this is deep enough that any such scenario will absolutely be detected.

File: bench/Micro.Benchmark/Micro.Benchmark.csproj [hunk 1]
```diff
@@ -16,6 +16,9 @@
     <PackageReference Include="BenchmarkDotNet.Diagnostics.Windows" Version="0.13.2" />
     <PackageReference Include="BenchmarkDotNet" Version="0.13.2" />
     <PackageReference Include="xunit" Version="2.4.2" />
+    <PackageReference Include="BenchmarkDotNet.Diagnostics.Windows" Version="0.13.1.1847" />
+    <PackageReference Include="BenchmarkDotNet" Version="0.13.1.1847" />
+    <PackageReference Include="xunit" Version="2.4.1" />
```
Comment (ppekrol) [inline bench/Micro.Benchmark/Micro.Benchmark.csproj:21]: merge issue

File: src/Corax/IndexWriter.cs [hunk 2]
```diff
@@ -1326,29 +1327,32 @@ private unsafe void InsertNumericFieldDoubles(Tree fieldsTree, int fieldId, Span
...
+                        continue; // we don't need to store duplicates
 
-                pos += ZigZagEncoding.Encode(tmpBuf, entry, pos);
-            }
+                    pos += ZigZagEncoding.Encode(tmpBufferPtr, entry, pos);
+                }
 
-            encoded = tmpBuf[..pos];
-            return true;
+                encoded = tmpBuf[..pos];
```
Comment (ppekrol) [inline src/Corax/IndexWriter.cs:1382]: just to make sure, the tmpBuf is our native memory? Asking if this is 'pinned'.

Comment (ayende) [reply src/Corax/IndexWriter.cs:1382]: This is unmanaged memory, which is allocated on `Commit`, no need to pin.

**Comment by arekpalinski** [comment]
@redknightlois Can you please solve the conflicts? I've tried to do it but wasn't sure so I'm leaving that to you

File: src/Sparrow.Server/Compression/Encoder3Gram.cs [hunk 3]
```diff
@@ -14,10 +14,14 @@ namespace Sparrow.Server.Compression
     [StructLayout(LayoutKind.Explicit)]
     internal unsafe struct Interval3Gram
     {
+        [FieldOffset(0)]
```
Comment (arekpalinski) [inline src/Sparrow.Server/Compression/Encoder3Gram.cs:17]: Just to make sure - this has `FieldOffset(0)` - same as `KeyBuffer`. When this field is used then we don't make use of `KeyBuffer`, right?

Comment (ayende) [reply src/Sparrow.Server/Compression/Encoder3Gram.cs:17]: This is meant to be a union, so you can refer to the same memory as either field.

**Pull request merged**